### ISSUE TITLE
Bizarre edge case with typelib defined byte() to long() implicit conversion

### ISSUE THREAD

**Issue by fafalone** [twinbasic/twinbasic#1689]
**Describe the bug**
VB allows this implicit conversion when typelib-defined aliases are substituted.

This is obviously improper and doesn't work:

```
Private Sub RcvLong(a() As Long)
ReDim a(0)
a(0) = 1
End Sub

Private Sub SndByte()
Dim b() As Byte
RcvLong b()
Debug.Print b(0)
End Sub
```

But with the following typelib aliases:

```
    typedef [public] int GLuint;

    typedef [public] unsigned char GLubyte;
```

The same thing *is* allowed in VB6

```
Private Sub RcvLong(a() As GLuint)
ReDim a(0)
a(0) = 1
End Sub

Private Sub SndByte()
Dim b() As GLbyte
RcvLong b()
Debug.Print b(0)
End Sub
```

Even though you'd think VB would just see this the same as the Byte and Long.

tB does not currently allow this.

**To Reproduce**
The above code in a project with vbogl.tlb.

**Expected behavior**
We may not *like* that this kind of voodoo is allowed, but tB will unfortunately have to support it.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version 0.15.398

**Additional context**
I think it's worth some discussion about how to handle this. Obviously the TLB scenario has to be supported for compatibility. But tB plans to have locally defined aliases too; will this be allowed with those? Should it be allowed with the native types?

vbogl.tlb attached here again for convenience.
[vbogl.zip](https://github.com/twinbasic/twinbasic/files/12774025/vbogl.zip)

**Comment by WaynePhillipsEA** [comment]
This was "fixed" to allow this edge case through the compilation stages in BETA 441, but this has highlighted a discrepancy in how VBx and tB access the elements in a safearray.  Let's take this example:

```
Private Sub RcvLong(a() As GLuint)
ReDim a(4)
a(0) = 1
a(1) = 2
a(2) = 3
a(3) = 4
a(4) = 5
End Sub

Private Sub SndByte()
Dim b() As GLubyte
RcvLong b()
Debug.Print b(0)
Debug.Print b(1)
Debug.Print b(2)
Debug.Print b(3)
Debug.Print b(4)
End Sub
```

VBx outputs `1`,`0`,`0`,`0`,`2`.
tB outputs `1`,`2`,`3`,`4`,`5`.

The actual SAFEARRAY definition and content matches for both VBx and tB, in that `b` contains a SAFEARRAY of 5 longs, but the mismatching definition, with `b` being declared in code as an array of bytes, causes a difference here.   

VBx is accessing the elements by computing the physical offsets using the **compile-time** size of elements (being 1 byte), and then accessing the byte at that offset.

tB is accessing the elements by computing the physical offsets using the **runtime** size of elements as stored in the SAFEARRAY cbElements field, and then reading the byte at that offset (skipping over the 3 bytes in-between each element).

Honestly, this edge case is nuts and shouldn't have ever been allowed in VBx.  Allowing the runtime SAFEARRAY structure to not match the defined array in code is spectacularly awful.  Neither implementation is "right" since there is no "right" here at all.

**Comment by WaynePhillipsEA** [comment]
BTW we can easily check that only the 'Byte' is being read:
```
Private Sub RcvLong(a() As GLuint)
ReDim a(4)
a(0) = 1 + 256
a(1) = 2 + 256
a(2) = 3 + 256
a(3) = 4 + 256
a(4) = 5 + 256
End Sub

Private Sub SndByte()
Dim b() As GLubyte
Dim c As Long
RcvLong b()
Debug.Print b(0)
Debug.Print b(1)
Debug.Print b(2)
Debug.Print b(3)
Debug.Print b(4)
End Sub
```

The results are the same as before, only the lower 'Byte' as declared at compile-time is being read (both VBx and tB).

**Comment by WaynePhillipsEA** [comment]
I will get tB fixed up so that we use the compile-time size of elements, and then we should be fully matching VBx behaviors here.

**Comment by WaynePhillipsEA** [comment]
My suspicion about reference indexing in VBx being difference can be confirmed:

```
Private Sub RcvLong(a() As GLuint)
ReDim a(4)
a(0) = 1 '+ 256
a(1) = 2 '+ 256
a(2) = 3 '+ 256
a(3) = 4 '+ 256
a(4) = 5 '+ 256
End Sub

Private Sub SndByte()
Dim b() As GLubyte
Dim c As Long
RcvLong b()
TestVal b(0)
TestVal b(1)
TestVal b(2)
TestVal b(3)
TestVal b(4)
End Sub

Private Sub TestVal(ByRef something As Byte)
    Debug.Print something
End Sub
```

`ByRef something` gives the same results as tB (offsets calculated with the SAFEARRAY runtime element size).  Changeing to `ByVal something`, you can see VBx calculates the offset using the compile-time element size.

**Comment by bclothier** [comment]
Sounds to me this merits multiple warnings given different ways to shoot one foot's off. I shudder to think about what if I had to maintain such project that relied on this zany implementation....

**Comment by WaynePhillipsEA** [comment]
Just to add more inconsistencies... VBx pcode implementation differs to native code here too.   Native code looks to stick with the compile-time size of elements, only the pcode/IDE implementation seems to use the runtime approach (and only when a reference to the element is taken).

**Comment by WaynePhillipsEA** [comment]
I'm somewhat inclined to just revert the change and disallow this edge case, given the inconsistent behavior seen in VBx.

**Comment by bclothier** [comment]
Just to consider - I understand that in some cases, VBx will play pretend that it's working with C-style array particularly with declare statements. Could this be the case here?

**Comment by WaynePhillipsEA** [comment]
No, it's very much SAFEARRAY handling.   e.g. the safearray bounds are checked, although against nonsensical values due to the mismatching subtypes.   If there's code out there relying on such quirky implementation behavior, then it deserves to blow up IMO.

**Comment by bclothier** [comment]
In this case, even if we disallow it, it might be a good idea to make it clear that this is disallowed due to VBx bug in the message so that it's communicated clearly and cuts off "but it works in VBx!"

**Comment by WaynePhillipsEA** [comment]
I'm going to allow the behavior when calling out to a TLB defined DLL declare (but output the TB0017 warning).   For internal use, I'm going to completely disallow it with a full compile error.  Good call - I will add a note that it's disallowed on purpose.

**Comment by fafalone** [comment]
If it helps, here's the real world example of this edge case being used. It's one of NeHe's very popular OpenGL tutorials, which is bad because this isn't even some ultra-obscure proof of concept, it's been used by tons of VB programs.

[Lesson6.zip](https://github.com/twinbasic/twinbasic/files/14223284/Lesson6.zip)

**Comment by WaynePhillipsEA** [comment]
Sigh.  I will hold off on any further changes for now then.  We'll probably have to adjust tB to match the general case where the offsets are calculated using the compile-time defined element sizes, which seems to match VBx **_most_** of the time.   This is going to need a lot of tests though, so will have to wait a while.

**Comment by WaynePhillipsEA** [comment]
> What's going on here? Is there an error check pass that doesn't resolve aliases, and VB6 is just trusting the issue didn't make it through that?

Yup, pretty much that.  tB has a similar 'verification' step, but we delve deeper when encountering alias subtypes.
